experimental/unicodeset_parser/src/parse.rs

```diff
@@ -336,149 +482,294 @@ where
+                let (offset, var_or_anchor) = self.parse_variable()?;
+                match var_or_anchor {
+                    VarOrAnchor::A => Ok((offset, false, MainToken::DollarSign)),
+                    VarOrAnchor::V(&VariableValue::Char(c)) => {
+                        Ok((offset, true, MainToken::CharOrString(c.into())))
+                    }
+                    VarOrAnchor::V(VariableValue::String(s)) => {
+                        // .into() handles 1-length-string -> char conversion
+                        Ok((offset, true, MainToken::CharOrString(s.clone().into())))
+                    }
+                    VarOrAnchor::V(VariableValue::UnicodeSet(set)) => {
+                        Ok((offset, true, MainToken::UnicodeSet(set.clone())))
+                    }
+                }
             }
+            // string
+            ('{', _) => self
+                .parse_string()
+                .map(|(offset, cs)| (offset, false, MainToken::CharOrString(cs))),
+            // inner set
+            ('\\', 'p' | 'P') | ('[', _) => {
+                let mut inner_builder = UnicodeSetBuilder::new_internal(
+                    self.iter,
+                    self.variable_map,
+                    // zero-copy clone
+                    self.xid_start,
+                    // zero-copy clone
+                    self.xid_continue,
+                    self.property_provider,
+                );
+                inner_builder.parse_unicode_set()?;
+                let (single, string_set) = inner_builder.finalize();
+                // note: offset - 1, because we already consumed full set
+                let offset = self.must_peek_index()? - 1;
+                let mut strings = string_set.into_iter().collect::<Vec<_>>();
+                strings.sort();
+                let cpilasl = CodePointInversionListAndStringList::try_from(
+                    single.build(),
+                    VarZeroVec::from(&strings),
+                )
+                .map_err(|_| PEK::Internal.with_offset(offset))?;
+                Ok((offset, false, MainToken::UnicodeSet(cpilasl)))
+            }
+            (c, _) if legal_char_start(c) => self
+                .parse_char()
+                .map(|(offset, cs)| (offset, false, MainToken::CharOrString(cs))),
+            ('-', _) => {
+                self.iter.next();
+                Ok((initial_offset, false, MainToken::Minus))
+            }
+            ('&', _) => {
+                self.iter.next();
+                Ok((initial_offset, false, MainToken::Ampersand))
+            }
+            (c, _) => Err(PEK::UnexpectedChar(c).with_offset(initial_offset)),
+        }
+    }
+
+    // parses a variable or an anchor. expects '$' as next token.
+    // is 'context-sensitive' to avoid duplicate work
+    // if this is a trailing $ (eg [.... $ ]), then this function returns Ok(Some((offset, VarOrAnchor::A)))
+    fn parse_variable(&mut self) -> Result<(usize, VarOrAnchor<'a>)> {
```

Comment:
`VarOrAnchor<'a>` could be `Option<&'a VariableValue<'a>>`

Reply:
Thanks! Wanted to do that but forgot